core/src/main/java/org/mule/runtime/core/api/strategy/factory/NonBlockingProcessingStrategyFactory.java

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.core.api.strategy.factory;
+
+import static java.util.concurrent.Executors.newFixedThreadPool;
+import static reactor.core.publisher.Flux.from;
+import static reactor.core.scheduler.Schedulers.fromExecutorService;
+
+import org.mule.runtime.api.exception.MuleException;
+import org.mule.runtime.core.api.Event;
+import org.mule.runtime.core.api.MuleContext;
+import org.mule.runtime.core.api.lifecycle.Startable;
+import org.mule.runtime.core.api.lifecycle.Stoppable;
+import org.mule.runtime.core.api.processor.MessageProcessorChainBuilder;
+import org.mule.runtime.core.api.processor.NonBlockingMessageProcessor;
+import org.mule.runtime.core.api.processor.ProcessingStrategy;
+import org.mule.runtime.core.api.processor.Processor;
+import org.mule.runtime.core.processor.strategy.AbstractThreadingProfileProcessingStrategy;
+
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.function.Function;
+
+import org.reactivestreams.Publisher;
+
+/**
+ * Allows Mule to use non-blocking execution model where possible and free up threads when performing IO operations.
+ *
+ * @since 3.7
+ */
+public class NonBlockingProcessingStrategyFactory implements ProcessingStrategyFactory {
+
+  protected Integer maxThreads;
+  protected Integer minThreads;
+  protected Integer maxBufferSize;
+  protected Long threadTTL;
+  protected Long threadWaitTimeout;
+  protected Integer poolExhaustedAction;
+
+  @Override
+  public ProcessingStrategy create() {
+    final NonBlockingProcessingStrategy processingStrategy = new NonBlockingProcessingStrategy();
+
+    if (maxThreads != null) {
+      processingStrategy.setMaxThreads(maxThreads);
+    }
+    if (minThreads != null) {
+      processingStrategy.setMinThreads(minThreads);
+    }
+    if (maxBufferSize != null) {
+      processingStrategy.setMaxBufferSize(maxBufferSize);
+    }
+    if (threadTTL != null) {
+      processingStrategy.setThreadTTL(threadTTL);
+    }
+    if (threadWaitTimeout != null) {
+      processingStrategy.setThreadWaitTimeout(threadWaitTimeout);
+    }
+    if (poolExhaustedAction != null) {
+      processingStrategy.setPoolExhaustedAction(poolExhaustedAction);
+    }
+
+    return processingStrategy;
+  }
+
+  // TODO MULE-10544 use internal executor, do not allow it to be passed form outside
```

Comment:
Need to take into account the ability to test still though.  Maybe using a TestScheduler if not via this means?